tests/default_vals.rs

```diff
@@ -411,23 +453,42 @@ fn conditional_reqs_pass() {
         .version("1.0")
         .author("F0x06")
         .about("Arg test")
-        .arg(Arg::with_name("target")
-            .takes_value(true)
-            .default_value("file")
-            .possible_values(&["file", "stdout"])
-            .long("target"))
-        .arg(Arg::with_name("input")
-            .takes_value(true)
-            .required(true)
-            .long("input"))
-        .arg(Arg::with_name("output")
-            .takes_value(true)
-            .required_if("target", "file")
-            .long("output"))
+        .arg(
+            Arg::with_name("target")
+                .takes_value(true)
+                .default_value("file")
+                .possible_values(&["file", "stdout"])
+                .long("target"),
+        )
+        .arg(
+            Arg::with_name("input")
+                .takes_value(true)
+                .required(true)
+                .long("input"),
+        )
+        .arg(
+            Arg::with_name("output")
+                .takes_value(true)
+                .required_if("target", "file")
+                .long("output"),
+        )
         .get_matches_from_safe(vec!["test", "--input", "some", "--output", "other"]);
 
     assert!(m.is_ok());
     let m = m.unwrap();
     assert_eq!(m.value_of("output"), Some("other"));
     assert_eq!(m.value_of("input"), Some("some"));
 }
+
+#[test]
```

Comment:
additional test I'll be adding: validate order when default is present and no arg. let me know if there are others you think are needed...

Reply:
My biggest concern for testing is to ensure correct order (from highest to lowest):

* Args from command line
* ENV vars
* Default Values

Also a test to confirm something along the lines of:

`$ MYVAR='some' program --option`

Where `--option` accepts 0 or one value. In this case, `matches.value_of("option") == None` because the empty value was specified at the command line (same for `--option=` and `--option ""`).

The only one I'm a little unsure about is:

`$ MYVAR='some' program --option`

Where `--option` accepts 0 or **more** values (i.e. multiples). Should the result be `["", "some"]` for the values, or `None`? The same goes for `$ MYVAR='some' program --option=other`. Should it be `["other", "some"]` or `["other"]`?

I'm actually fine with either way, so long as it's documented which way it'll happen. I'm inclined to lean specified args (at the command line) override env vars entirely, so the examples above would result in `None` and `["other"]` respectively. Otherwise there is no way to "override" ENV vars without manually resetting them `$ MYVAR="" program --option` which would be a pain.

The case for "merging" env vars and command line options would be more in line with allowing external argv's as in `$ MYARGV="--option some" program --option other`. But that's a whole other issue :stuck_out_tongue_winking_eye: so let's not worry about it here.

Reply:
I'd be concerned with merging as opposed to overriding. I suppose we could allow that through some other mechanism. But I think it would get confusing as to how to maintain that from a user perspective.

As to multi-opts, I'll through in as many test cases as I can and we'll see how far we get...

Reply:
Yeah I don't want to mess with merging right now. I'd like to stick with just overriding :smile: